File: patcha/collectors/git.py
import json
import logging
from collections import defaultdict
from datetime import datetime, timezone
from pathlib import Path
from typing import List, Optional
import git
from git import Repo

from patcha.config import config
from patcha.db.models import Event, EventType, GitCommit, GitStash
from patcha.utils.guard import CollectorGuard
from patcha.utils.jsonl import trim_jsonl

logger = logging.getLogger(__name__)

# Staging snapshots carry full diffs and are appended every poll, so keep a
# tighter bound than the lightweight window/screen logs.
_STAGE_MAX_LOG_ROWS = 20_000
_STAGE_TRIM_EVERY = 500


class GitCollector:
    def __init__(self, repo_path: Optional[str] = None):
        self.repo_path = Path(repo_path) if repo_path else Path.cwd()
        self._commits_guard = CollectorGuard("git_commits")
        self._stashes_guard = CollectorGuard("git_stashes")
        self._stage_write_count: int = 0

    def is_git_repo(self) -> bool:
        try:
            Repo(self.repo_path)
            return True
        except git.exc.InvalidGitRepositoryError:
            return False

    _SKIP_DIRS = {
        "Music",
        "Pictures",
        "Movies",
        "Library",
        "Applications",
        "iCloud Drive",
        "iCloud",
    }

    def _find_git_repos(self, search_path: Optional[Path] = None) -> List[Path]:
        search_path = search_path or self.repo_path
        git_repos: List[Path] = []

        def _walk(path: Path, depth: int) -> None:
            if depth > 3:
                return
            try:
                for entry in path.iterdir():
                    if not entry.is_dir():
                        continue
                    name = entry.name
                    if name.startswith("."):
                        continue
                    if name in self._SKIP_DIRS:
                        continue
                    if name == ".git":
                        if path not in git_repos:
                            git_repos.append(path)
                        return
                    _walk(entry, depth + 1)
            except PermissionError:
                pass
            except Exception as e:
                logger.debug("Error walking %s: %s", path, e)

        try:
            if (search_path / ".git").exists():
                git_repos.append(search_path)
            else:
                _walk(search_path, 1)
        except Exception as e:
            logger.debug("Error searching for git repositories: %s", e)

        return git_repos

    def collect_commits(self, since: Optional[datetime] = None) -> List[Event]:
        if not self._commits_guard.ok:
            return []
        events = []

        git_repos = self._find_git_repos(Path.home())
        if not git_repos:
            return []

        for repo_path in git_repos:
            try:
                repo = Repo(repo_path)

                # If since is provided, only get commits after that time
                if since:
                    # Use git log with --since parameter for efficiency
                    commits = list(
                        repo.iter_commits(since=since.strftime("%Y-%m-%d %H:%M:%S"))
                    )
                else:
                    commits = list(
                        repo.iter_commits(max_count=50)
                    )  # Limit to recent 50 commits for performance

                for commit in commits:
                    commit_time = datetime.fromtimestamp(
                        commit.committed_date, tz=timezone.utc
                    )

                    # Double-check the timestamp (since git --since might be inclusive)
                    if since and commit_time < since:
                        continue

                    all_files_changed = list(commit.stats.files.keys())

                    try:
                        branch = repo.active_branch.name
                    except TypeError:
                        branch = "unknown"

                    try:
                        if commit.parents:
                            diff_text = repo.git.diff(
                                commit.parents[0].hexsha, commit.hexsha
                            )
                        else:
                            diff_text = ""
                    except Exception:
                        diff_text = ""

                    git_commit = GitCommit(
                        hash=commit.hexsha,
                        message=commit.message.strip(),
                        author=str(commit.author),
                        timestamp=commit_time,
                        files_changed=all_files_changed[:30],
                        insertions=commit.stats.total["insertions"],
                        deletions=commit.stats.total["deletions"],
                        branch=branch,
                        diff=diff_text,
                    )

                    event = Event(
                        timestamp=commit_time,
                        type=EventType.GIT_COMMIT,
                        source="git",
                        project=repo_path.name,
                        raw_content=json.dumps(git_commit.model_dump(), default=str),
                        source_doc_id=f"{repo_path}:{commit.hexsha}",
                        metadata={
                            "repo_path": str(repo_path),
                            "files_count": len(all_files_changed),
                            "lines_changed": git_commit.insertions
                            + git_commit.deletions,
                        },
                    )
                    events.append(event)
                self._commits_guard.success()

            except Exception as e:
                self._commits_guard.fail(e, msg=f"{repo_path}: {e}")
                continue

        return events

    def collect_stashes(self, since: Optional[datetime] = None) -> List[Event]:
        if not self._stashes_guard.ok or not self.is_git_repo():
            return []

        events = []
        try:
            repo = Repo(self.repo_path)

            stash_lines = repo.git.stash(
                "list", "--format=%gd%x00%H%x00%ct%x00%gs"
            ).split("\n")
            for stash in stash_lines:
                if not stash.strip():
                    continue

                stash_parts = stash.split("\x00", 3)
                if len(stash_parts) != 4:
                    continue

                stash_name, stash_hash, stash_timestamp, stash_message = stash_parts
                try:
                    stash_time = datetime.fromtimestamp(
                        int(stash_timestamp), tz=timezone.utc
                    )
                except ValueError:
                    stash_time = datetime.now(timezone.utc)

                if since and stash_time < since:
                    continue

                try:
                    stash_info = repo.git.stash("show", "--stat", stash_name)
                    files_changed = [
                        line.split("|")[0].strip()
                        for line in stash_info.split("\n")[:-1]
                    ]
                except Exception:
                    files_changed = []

                git_stash = GitStash(
                    name=stash_name,
                    message=stash_message,
                    timestamp=stash_time,
                    files_changed=files_changed,
                )

                event = Event(
                    timestamp=git_stash.timestamp,
                    type=EventType.GIT_STASH,
                    source="git",
                    project=self._get_project_name(),
                    raw_content=json.dumps(git_stash.model_dump(), default=str),
                    source_doc_id=f"{self.repo_path.resolve()}:stash:{stash_hash}",
                    metadata={
                        "repo_path": str(self.repo_path),
                        "stash_hash": stash_hash,
                        "files_count": len(files_changed),
                    },
                )
                events.append(event)

            self._stashes_guard.success()
        except Exception as e:
            self._stashes_guard.fail(e)

        return events

    def _get_project_name(self, repo_path: Optional[Path] = None) -> str:
        target_path = repo_path or self.repo_path
        return target_path.name

    def record_staging_snapshot(self) -> None:
        """Snapshot staged/unstaged state for all known repos and append to log."""
        repos = self._find_git_repos(Path.home())
        config.data_dir.mkdir(parents=True, exist_ok=True)
        snapshot_file = config.data_dir / "git_stage_snapshots.jsonl"

        for repo_path in repos:
            try:
                repo = Repo(repo_path)

                try:
                    staged = [d.a_path for d in repo.index.diff("HEAD")]
                except Exception:
                    staged = []

                unstaged = [d.a_path for d in repo.index.diff(None)]
                untracked = list(repo.untracked_files)[:20]

                try:
                    staged_diff = repo.git.diff("HEAD", "--staged")
                except Exception:
                    staged_diff = ""

                entry = {
                    "timestamp": datetime.now(timezone.utc).isoformat(),
                    "repo": str(repo_path),
                    "staged": staged,
                    "unstaged": unstaged,
                    "untracked": untracked,
                    "staged_diff": staged_diff,
                }
                with open(snapshot_file, "a") as f:
                    f.write(json.dumps(entry) + "\n")

                self._stage_write_count += 1
                if self._stage_write_count % _STAGE_TRIM_EVERY == 0:
                    trim_jsonl(snapshot_file, _STAGE_MAX_LOG_ROWS)

            except Exception:
                continue

    def collect_staging_events(self, since: datetime) -> List[Event]:
        """Return events for when the staged file set changed since `since`.

        Emits one event per change in staged files — e.g. when the user
        runs `git add` or `git reset`. Consecutive snapshots with identical
        staged files produce no event.
        """
        snapshot_file = config.data_dir / "git_stage_snapshots.jsonl"
        if not snapshot_file.exists():
            return []

        snapshots_by_repo: dict = defaultdict(list)
        with open(snapshot_file) as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue
                try:
                    entry = json.loads(line)
                    ts = datetime.fromisoformat(entry["timestamp"])
                    if ts >= since:
                        snapshots_by_repo[entry["repo"]].append((ts, entry))
                except (json.JSONDecodeError, KeyError, ValueError):
                    continue

        events: List[Event] = []

        for repo, snapshots in snapshots_by_repo.items():
            if len(snapshots) < 2:
                continue

            prev_staged = set(snapshots[0][1].get("staged", []))

            for ts, entry in snapshots[1:]:
                curr_staged = set(entry.get("staged", []))
                newly_staged = curr_staged - prev_staged
                newly_unstaged = prev_staged - curr_staged

                if newly_staged or newly_unstaged:
                    project = Path(repo).name
                    all_staged = sorted(curr_staged)
                    file_list = ", ".join(all_staged[:10])
                    suffix = (
                        f" (+{len(all_staged) - 10} more)"
                        if len(all_staged) > 10
                        else ""
                    )

                    parts = []
                    if newly_staged:
                        parts.append(f"staged: {', '.join(sorted(newly_staged))}")
                    if newly_unstaged:
                        parts.append(f"unstaged: {', '.join(sorted(newly_unstaged))}")

                    diff_content = entry.get("staged_diff", "")
                    diff_section = f"\nDiff:\n{diff_content}" if diff_content else ""

                    raw = (
                        f"Git index change in {project}\n"
                        f"Change: {'; '.join(parts)}\n"
                        f"Currently staged: {file_list}{suffix}"
                        f"{diff_section}"
                    )

                    events.append(
                        Event(
                            timestamp=ts,
                            type=EventType.GIT_STAGED,
                            source="git",
                            project=project,
                            raw_content=raw,
                            metadata={
                                "repo_path": repo,
                                "staged_files": all_staged,
                                "newly_staged": sorted(newly_staged),
                                "newly_unstaged": sorted(newly_unstaged),
                                "unstaged_files": entry.get("unstaged", []),
                            },
                        )
                    )

                prev_staged = curr_staged

        return events

File: rust/patcha/src/collectors/git.rs
use crate::{
    collectors::guard::CollectorGuard,
    models::{Event, EventType},
};
use anyhow::Result;
use chrono::{DateTime, Utc};
use git2::{Repository, Sort};
use std::collections::{HashMap, HashSet};
use std::io::Write;
use std::path::{Path, PathBuf};

const STAGE_MAX_LOG_ROWS: usize = 20_000;
const STAGE_TRIM_EVERY: usize = 500;

// Directories skipped when scanning for git repos (same as Python)
const SKIP_DIRS: &[&str] = &[
    "Music", "Pictures", "Movies", "Library",
    "Applications", "iCloud Drive", "iCloud",
];

pub struct GitCollector {
    commits_guard: CollectorGuard,
    stashes_guard: CollectorGuard,
    data_dir: PathBuf,
    stage_write_count: usize,
}

impl GitCollector {
    pub fn new(data_dir: PathBuf) -> Self {
        Self {
            commits_guard: CollectorGuard::new("git_commits", 3),
            stashes_guard: CollectorGuard::new("git_stashes", 3),
            data_dir,
            stage_write_count: 0,
        }
    }

    // -----------------------------------------------------------------------
    // Commits
    // -----------------------------------------------------------------------

    pub fn collect_commits(&mut self, since: Option<DateTime<Utc>>) -> Vec<Event> {
        if !self.commits_guard.ok() {
            return Vec::new();
        }
        match self.collect_commits_inner(since) {
            Ok(events) => {
                self.commits_guard.success();
                events
            }
            Err(e) => {
                self.commits_guard.fail(&e);
                Vec::new()
            }
        }
    }

    fn collect_commits_inner(&self, since: Option<DateTime<Utc>>) -> Result<Vec<Event>> {
        let repos = find_git_repos(&dirs::home_dir().unwrap_or_default());
        let mut events = Vec::new();

        for repo_path in repos {
            let Ok(repo) = Repository::open(&repo_path) else {
                continue;
            };

            let mut revwalk = match repo.revwalk() {
                Ok(rw) => rw,
                Err(_) => continue,
            };
            let _ = revwalk.push_head();
            revwalk.set_sorting(Sort::TIME)?;

            let limit = if since.is_none() { 50 } else { usize::MAX };
            let mut count = 0;

            for oid in revwalk.flatten() {
                if count >= limit {
                    break;
                }
                let Ok(commit) = repo.find_commit(oid) else {
                    continue;
                };

                let commit_ts = commit.time().seconds();
                let timestamp =
                    DateTime::from_timestamp(commit_ts, 0).unwrap_or_else(Utc::now);

                if let Some(since) = since {
                    if timestamp < since {
                        break;
                    }
                }

                let message = commit.message().unwrap_or("").trim().to_owned();
                let author = commit.author().name().unwrap_or("").to_owned();
                let hash = oid.to_string();

                let branch = repo
                    .head()
                    .ok()
                    .and_then(|h| h.shorthand().map(|s| s.to_owned()))
                    .unwrap_or_else(|| "unknown".into());

                // Get diff stats
                let (files_changed, insertions, deletions, diff_text) =
                    commit_stats(&repo, &commit);

                let raw = serde_json::json!({
                    "hash": hash,
                    "message": message,
                    "author": author,
                    "timestamp": timestamp.to_rfc3339(),
                    "files_changed": files_changed,
                    "insertions": insertions,
                    "deletions": deletions,
                    "branch": branch,
                    "diff": diff_text,
                });

                let mut e = Event::new(EventType::GitCommit, raw.to_string());
                e.timestamp = timestamp;
                e.source = Some("git".to_owned());
                e.project = Some(repo_path.file_name().unwrap_or_default().to_string_lossy().into_owned());
                e.source_doc_id = Some(format!("{}:{}", repo_path.display(), hash));
                e.metadata.insert("repo_path".into(), serde_json::json!(repo_path.to_string_lossy()));
                e.metadata.insert("files_count".into(), serde_json::json!(files_changed.len()));
                e.metadata.insert("lines_changed".into(), serde_json::json!(insertions + deletions));
                e.metadata.insert("files_changed".into(), serde_json::json!(files_changed));

                events.push(e);
                count += 1;
            }
        }

        Ok(events)
    }

    // -----------------------------------------------------------------------
    // Staging snapshots + events
    // -----------------------------------------------------------------------

    /// Snapshot the staged/unstaged state for all repos and append to log.
    pub fn record_staging_snapshot(&mut self) {
        let repos = find_git_repos(&dirs::home_dir().unwrap_or_default());
        let snapshot_file = self.data_dir.join("git_stage_snapshots.jsonl");

        if let Some(parent) = snapshot_file.parent() {
            let _ = std::fs::create_dir_all(parent);
        }

        for repo_path in repos {
            let Ok(repo) = Repository::open(&repo_path) else {
                continue;
            };

            let (staged, unstaged, untracked, staged_diff) = staging_state(&repo);

            let entry = serde_json::json!({
                "timestamp": Utc::now().to_rfc3339(),
                "repo": repo_path.to_string_lossy(),
                "staged": staged,
                "unstaged": unstaged,
                "untracked": untracked,
                "staged_diff": staged_diff,
            });

            if let Ok(mut f) = std::fs::OpenOptions::new()
                .create(true)
                .append(true)
                .open(&snapshot_file)
            {
                let _ = writeln!(f, "{}", entry);
            }

            self.stage_write_count += 1;
            if self.stage_write_count % STAGE_TRIM_EVERY == 0 {
                let _ = trim_jsonl(&snapshot_file, STAGE_MAX_LOG_ROWS);
            }
        }
    }

    /// Emit events for changes in the staged file set since `since`.
    pub fn collect_staging_events(&self, since: DateTime<Utc>) -> Vec<Event> {
        let snapshot_file = self.data_dir.join("git_stage_snapshots.jsonl");
        if !snapshot_file.exists() {
            return Vec::new();
        }

        let Ok(content) = std::fs::read_to_string(&snapshot_file) else {
            return Vec::new();
        };

        // Group snapshots by repo
        let mut by_repo: HashMap<String, Vec<(DateTime<Utc>, serde_json::Value)>> =
            HashMap::new();

        for line in content.lines() {
            let Ok(obj) = serde_json::from_str::<serde_json::Value>(line) else {
                continue;
            };
            let Some(ts_str) = obj.get("timestamp").and_then(|v| v.as_str()) else {
                continue;
            };
            let Ok(ts) = DateTime::parse_from_rfc3339(ts_str) else {
                continue;
            };
            let ts = ts.with_timezone(&Utc);
            if ts < since {
                continue;
            }
            let repo = obj.get("repo").and_then(|v| v.as_str()).unwrap_or("").to_owned();
            by_repo.entry(repo).or_default().push((ts, obj));
        }

        let mut events = Vec::new();

        for (repo, snapshots) in &by_repo {
            if snapshots.len() < 2 {
                continue;
            }

            let mut prev_staged: HashSet<String> = snapshot_files(&snapshots[0].1, "staged");

            for (ts, entry) in &snapshots[1..] {
                let curr_staged: HashSet<String> = snapshot_files(entry, "staged");
                let newly_staged: HashSet<_> = curr_staged.difference(&prev_staged).collect();
                let newly_unstaged: HashSet<_> = prev_staged.difference(&curr_staged).collect();

                if newly_staged.is_empty() && newly_unstaged.is_empty() {
                    prev_staged = curr_staged;
                    continue;
                }

                let project = Path::new(repo)
                    .file_name()
                    .map(|n| n.to_string_lossy().into_owned())
                    .unwrap_or_default();

                let all_staged: Vec<String> = {
                    let mut v: Vec<String> = curr_staged.iter().cloned().collect();
                    v.sort();
                    v
                };

                let file_list = all_staged[..all_staged.len().min(10)].join(", ");
                let suffix = if all_staged.len() > 10 {
                    format!(" (+{} more)", all_staged.len() - 10)
                } else {
                    String::new()
                };

                let mut parts = Vec::new();
                if !newly_staged.is_empty() {
                    let mut ns: Vec<_> = newly_staged.iter().collect();
                    ns.sort();
                    parts.push(format!("staged: {}", ns.iter().map(|s| s.as_str()).collect::<Vec<_>>().join(", ")));
                }
                if !newly_unstaged.is_empty() {
                    let mut nu: Vec<_> = newly_unstaged.iter().collect();
                    nu.sort();
                    parts.push(format!("unstaged: {}", nu.iter().map(|s| s.as_str()).collect::<Vec<_>>().join(", ")));
                }

                let diff = entry.get("staged_diff").and_then(|v| v.as_str()).unwrap_or("");
                let diff_section = if diff.is_empty() {
                    String::new()
                } else {
                    format!("\nDiff:\n{diff}")
                };

                let raw = format!(
                    "Git index change in {project}\nChange: {}\nCurrently staged: {file_list}{suffix}{diff_section}",
                    parts.join("; ")
                );

                let mut e = Event::new(EventType::GitStaged, raw);
                e.timestamp = *ts;
                e.source = Some("git".to_owned());
                e.project = Some(project);
                e.metadata.insert("repo_path".into(), serde_json::json!(repo));
                e.metadata.insert("staged_files".into(), serde_json::json!(all_staged));
                e.metadata.insert("newly_staged".into(), serde_json::json!(newly_staged.iter().collect::<Vec<_>>()));
                e.metadata.insert("newly_unstaged".into(), serde_json::json!(newly_unstaged.iter().collect::<Vec<_>>()));

                events.push(e);
                prev_staged = curr_staged;
            }
        }

        events
    }
}

// ---------------------------------------------------------------------------
// Repo discovery (max depth 3, skips known-slow dirs)
// ---------------------------------------------------------------------------

fn find_git_repos(search_path: &Path) -> Vec<PathBuf> {
    let mut repos = Vec::new();
    walk_for_repos(search_path, 0, &mut repos);
    repos
}

fn walk_for_repos(path: &Path, depth: usize, out: &mut Vec<PathBuf>) {
    if depth > 3 {
        return;
    }
    let Ok(entries) = std::fs::read_dir(path) else {
        return;
    };
    for entry in entries.flatten() {
        let child = entry.path();
        if !child.is_dir() {
            continue;
        }
        let name = child.file_name().and_then(|n| n.to_str()).unwrap_or("");
        if name.starts_with('.') || SKIP_DIRS.contains(&name) {
            continue;
        }
        if name == ".git" {
            out.push(path.to_path_buf());
            return;
        }
        walk_for_repos(&child, depth + 1, out);
    }
}

// ---------------------------------------------------------------------------
// git2 helpers
// ---------------------------------------------------------------------------

fn commit_stats(
    repo: &Repository,
    commit: &git2::Commit,
) -> (Vec<String>, usize, usize, String) {
    let mut files = Vec::new();
    let mut insertions = 0usize;
    let mut deletions = 0usize;
    let mut diff_text = String::new();

    let parent_tree = commit.parents().next().and_then(|p| p.tree().ok());
    let commit_tree = commit.tree().ok();

    if let (Some(ct), parent_opt) = (commit_tree, parent_tree) {
        if let Ok(diff) = repo.diff_tree_to_tree(parent_opt.as_ref(), Some(&ct), None) {
            let _ = diff.foreach(
                &mut |delta, _| {
                    if let Some(path) = delta.new_file().path() {
                        files.push(path.to_string_lossy().into_owned());
                    }
                    true
                },
                None,
                None,
                None,
            );

            if let Ok(stats) = diff.stats() {
                insertions = stats.insertions();
                deletions = stats.deletions();
            }

            let mut diff_bytes: Vec<u8> = Vec::new();
            if diff.print(git2::DiffFormat::Patch, |_d, _h, line| {
                diff_bytes.extend_from_slice(line.content());
                true
            }).is_ok() {
                diff_text = String::from_utf8_lossy(&diff_bytes).into_owned();
                if diff_text.len() > 16_000 {
                    diff_text.truncate(16_000);
                    diff_text.push_str("\n...(truncated)");
                }
            }
        }
    }

    files.truncate(30);
    (files, insertions, deletions, diff_text)
}

fn staging_state(repo: &Repository) -> (Vec<String>, Vec<String>, Vec<String>, String) {
    // Staged files: diff HEAD → index
    let staged = {
        let head_tree = repo
            .head()
            .ok()
            .and_then(|h| h.peel_to_tree().ok());
        let index = repo.index().ok();
        let mut files = Vec::new();
        if let (Some(tree), Some(idx)) = (head_tree, index) {
            if let Ok(diff) = repo.diff_tree_to_index(Some(&tree), Some(&idx), None) {
                let _ = diff.foreach(
                    &mut |delta, _| {
                        if let Some(p) = delta.new_file().path() {
                            files.push(p.to_string_lossy().into_owned());
                        }
                        true
                    },
                    None, None, None,
                );
            }
        }
        files
    };

    // Unstaged files: diff index → workdir
    let unstaged = {
        let mut files = Vec::new();
        if let Ok(diff) = repo.diff_index_to_workdir(None, None) {
            let _ = diff.foreach(
                &mut |delta, _| {
                    if let Some(p) = delta.old_file().path() {
                        files.push(p.to_string_lossy().into_owned());
                    }
                    true
                },
                None, None, None,
            );
        }
        files
    };

    // Untracked (max 20)
    let untracked: Vec<String> = {
        let mut opts = git2::StatusOptions::new();
        opts.include_untracked(true);
        repo.statuses(Some(&mut opts))
            .ok()
            .map(|statuses| {
                statuses
                    .iter()
                    .filter(|s| s.status().contains(git2::Status::WT_NEW))
                    .take(20)
                    .filter_map(|s| s.path().map(|p| p.to_owned()))
                    .collect()
            })
            .unwrap_or_default()
    };

    // Staged diff text
    let staged_diff = {
        let head_tree = repo.head().ok().and_then(|h| h.peel_to_tree().ok());
        let index = repo.index().ok();
        let mut diff_text = String::new();
        if let (Some(tree), Some(idx)) = (head_tree, index) {
            if let Ok(diff) = repo.diff_tree_to_index(Some(&tree), Some(&idx), None) {
                let mut buf = Vec::new();
                let _ = diff.print(git2::DiffFormat::Patch, |_, _, line| {
                    buf.extend_from_slice(line.content());
                    true
                });
                diff_text = String::from_utf8_lossy(&buf).into_owned();
                if diff_text.len() > 8_000 {
                    diff_text.truncate(8_000);
                }
            }
        }
        diff_text
    };

    (staged, unstaged, untracked, staged_diff)
}

fn snapshot_files(entry: &serde_json::Value, key: &str) -> HashSet<String> {
    entry
        .get(key)
        .and_then(|v| v.as_array())
        .map(|arr| arr.iter().filter_map(|v| v.as_str().map(|s| s.to_owned())).collect())
        .unwrap_or_default()
}

fn trim_jsonl(path: &Path, max_rows: usize) -> Result<()> {
    let content = std::fs::read_to_string(path)?;
    let lines: Vec<&str> = content.lines().collect();
    if lines.len() <= max_rows {
        return Ok(());
    }
    let keep = &lines[lines.len() - max_rows..];
    let tmp = path.with_extension("jsonl.tmp");
    std::fs::write(&tmp, keep.join("\n") + "\n")?;
    std::fs::rename(tmp, path)?;
    Ok(())
}
